hamster_gtk/misc/widgets/raw_fact_entry.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import absolute_import, unicode_literals
+import datetime
+
+from gi.repository import GObject, Gtk
+from orderedset import OrderedSet
+from six import text_type
+from hamster_lib import Category, Fact
+from hamster_gtk.helpers import _u
+from hamster_gtk import helpers
+
+
+class RawFactEntry(Gtk.Entry):
+    """A custom entry widgets that provides ``faw fact`` specific autocompletion behaviour."""
+
+    def __init__(self, controller, *args, **kwargs):
+        """Instantiate the class."""
+        super(RawFactEntry, self).__init__(*args, **kwargs)
+        self._controller = controller
+        self._controller.signal_handler.connect('facts-changed', self._on_facts_changed)
+        self.set_completion(RawFactCompletion(self._controller))
+
+    # Callbacks
+    def _on_facts_changed(self, evtl):
+        """Callback triggered when facts have changed."""
+        self.set_completion(RawFactCompletion(self._controller))
+
+
+class RawFactCompletion(Gtk.EntryCompletion):
+    """
+    Return a completion instance to match 'activity@category' strings.
+
+    Returns:
+        Gtk.EntryCompletion: Completion instance.
+    """
+
+    def __init__(self, controller, *args, **kwargs):
+        """instantiate class."""
+        super(RawFactCompletion, self).__init__(*args, **kwargs)
+        self._controller = controller
+        self.set_model(self._get_store())
+        self.set_text_column(0)
+        self.set_match_func(self._match_anywhere, None)
+        self.connect('match-selected', self._on_match_selected)
+
+    def _get_store(self):
+        store = Gtk.ListStore(GObject.TYPE_STRING, GObject.TYPE_STRING, GObject.TYPE_STRING)
+        for activity in self._get_activities():
+            category = ''
+            if activity.category:
+                category = text_type(activity.category.name)
+            store.append([
+                helpers.serialise_activity(activity), text_type(activity.name), category
+            ])
+        return store
+
+    def _get_activities(self):
+        today = datetime.date.today()
+        recent_activities = [fact.activity for fact in self._controller.facts.get_all(
+            start=today, end=today)]
+        return OrderedSet(recent_activities)
+
+    def _match_anywhere(self, entrystr, iter, data, *args):
+        """
+        Check if the entrysring is a substring of our text_column.
+
+        Args:
+            entrystr (str): The text extracted from the entry so far. Note that key is
+                normalized and case-folded (see GLib.utf8_normalize() and
+                GLib.utf8_casefold()).
+            iter: Iterator position indicating the model row we check against.
+            data: Arbitrary user data.
+
+        Returns:
+            boot: ``True`` if ``entrystring`` is a substring, ``False`` if not.
+
+        Note this is a custom match function, for details on the general generic
+        solution [please see|https://lazka.github.io/pgi-docs/#Gtk-3.0/
```

Comment:
Links should not really be broken like this. Also what is the `[link|url]` syntax?